src/api/layer.js

```diff
@@ -282,17 +282,15 @@ export default class Layer {
     }
     async _styleChanged(style) {
         await this._context;
-        const originalPromise = this.requestData(style);
-        if (!originalPromise) {
-            // The previous stored metadata is still valid
+        return this.requestMetadata(style).then(metadata => {
```

Comment:
Since this function is async you can use await

```javascript
 async _styleChanged(style) {
        await this._context;
        const metadata = await this.requestMetadata(style);
        if (metadata) {
            this.metadata = metadata;
            this._compileShaders(style, this.metadata);
            this.requestData();
        }
    }
```

Reply:
I'm not very sure about this, because _styleChanged requires to return a Promise as it is now.

Reply:
```javascript
// ... 
return this.requestData();
```

?

Reply:
Yes, I think `        const metadata = await this.requestMetadata(style);` is better.

Reply:
So with both suggestions the code changes to this one, right?

```js
async _styleChanged(style) {
    await this._context;
    this.metadata = await this.requestMetadata(style);
    this._compileShaders(style, this.metadata);
    return this.requestData();
}
```

Reply:
Maybe the last `return` could be removed, it depends on what kind of semantics do we want. Resolve the promise when the style is known to be valid vs resolve the promise when the style is known to be valid and required data have been downloaded.

I think the first one is better since the second one doesn't mean it has been painted yet, and we already have the `loaded` event.

In any case, the promise is mainly useful for tracking errors (`reject`), not so much for the `resolve`.